### PR TITLE
[10.0] add new module to manage sale promotion based on rules

### PR DESCRIPTION
**Pull request by lmignon**
A new repository 'sale_discount' could be a better place for this kind of addon.

### CODE REVIEW

File: sale_promotion_rule/__manifest__.py
Comment (sebastienbeau) [inline sale_promotion_rule/__manifest__.py]: Missing author "Acsone, Odoo Community Association (OCA)"
```suggestion
    "author": "Akretion, Acsone, Odoo Community Association (OCA)",
```

File: sale_promotion_rule/README.rst
Comment (sebastienbeau) [inline sale_promotion_rule/README.rst]: Missing OCA README info

**Review by sebastienbeau — APPROVED** [review]
Fonctionnal tested
Thanks for proposing in OCA.
Minor change to do regarding author and readme

File: sale_promotion_rule/README.rst
Comment (sebastienbeau) [inline sale_promotion_rule/README.rst:86]: you can add yourself (Laurent) as contributor

Comment (lmignon) [reply sale_promotion_rule/README.rst:86]: Thank you @sebastienbeau. Comments included in last commit.

**Review by Cedric-Pigeon — APPROVED** [review]

File: sale_promotion_rule/readme/DESCRIPTION.rst
Comment (legalsylvain) [inline sale_promotion_rule/readme/DESCRIPTION.rst]: You can now write : ../sale_promotion_rule/static/description/promotion_rule.png

(more simple, and make the url resilient, during the migration)

Not a blocking point.

Comment (lmignon) [reply sale_promotion_rule/readme/DESCRIPTION.rst]: thank you @legalsylvain Change done

File: sale_promotion_rule/__manifest__.py
Comment (rvalyi) [inline sale_promotion_rule/__manifest__.py]: this is strange: we have akretion in the website but not in the author

Comment (lmignon) [reply sale_promotion_rule/__manifest__.py]: I'll fix it... Should be Akretion in place of Anybox...

Comment (lmignon) [reply sale_promotion_rule/__manifest__.py]: @rvalyi fixed in https://github.com/OCA/sale-workflow/pull/729/commits/6fb82acd4c3538132d00a843acb58fe92afbb803

File: sale_promotion_rule/__manifest__.py
Comment (nikul-serpentcs) [inline sale_promotion_rule/__manifest__.py]: Replace URL as per [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L11)

File: sale_promotion_rule/__manifest__.py
Comment (nikul-serpentcs) [inline sale_promotion_rule/__manifest__.py]: @lmignon if not any `external_dependencies` then remove it

**Review by nikul-serpentcs — COMMENTED** [review]
Improve Code

File: sale_promotion_rule/__manifest__.py
Comment (lmignon) [reply sale_promotion_rule/__manifest__.py]: @nikul-serpentcs done

File: sale_promotion_rule/__manifest__.py
Comment (lmignon) [reply sale_promotion_rule/__manifest__.py]: @nikul-serpentcs done

File: sale_promotion_rule/models/sale_order.py
Comment (rousseldenis) [inline sale_promotion_rule/models/sale_order.py:52]: Even if each rule is cumulated, IMHO use of union is better

File: sale_promotion_rule/models/sale_promotion_rule.py
Comment (rousseldenis) [inline sale_promotion_rule/models/sale_promotion_rule.py]: IMHO the name is quite weird as we are waiting for a float. Maybe 'restriction_amount_filed ?

File: sale_promotion_rule/models/sale_promotion_rule.py
Comment (rousseldenis) [inline sale_promotion_rule/models/sale_promotion_rule.py]: Maybe a search_count is sufficent ?

File: sale_promotion_rule/models/sale_promotion_rule.py
Comment (rousseldenis) [inline sale_promotion_rule/models/sale_promotion_rule.py]: You should override name_get as display_name is automatically computed on it.

File: sale_promotion_rule/models/sale_promotion_rule.py
Comment (rousseldenis) [inline sale_promotion_rule/models/sale_promotion_rule.py:297]: browse() ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
LGTM. Some comments

File: sale_promotion_rule/models/sale_order.py
Comment (lmignon) [reply sale_promotion_rule/models/sale_order.py:52]: done

File: sale_promotion_rule/models/sale_promotion_rule.py
Comment (lmignon) [reply sale_promotion_rule/models/sale_promotion_rule.py]: done

File: sale_promotion_rule/models/sale_promotion_rule.py
Comment (lmignon) [reply sale_promotion_rule/models/sale_promotion_rule.py]: done

File: sale_promotion_rule/models/sale_promotion_rule.py
Comment (lmignon) [reply sale_promotion_rule/models/sale_promotion_rule.py]: done

File: sale_promotion_rule/models/sale_promotion_rule.py
Comment (lmignon) [reply sale_promotion_rule/models/sale_promotion_rule.py:297]: @rousseldenis arg of defaultdict must be a callable....

**Review by rousseldenis — APPROVED** [review]

**Comment by pedrobaeza** [comment]
@lmignon may you squash a bit commit history before merging?

**Comment by lmignon** [comment]
@pedrobaeza Sorry for the late reply, I didn't see your comment. It's a lot of work to rework/squash the commit history since I've 3 others PRs based on this one... 
https://github.com/OCA/sale-workflow/pulls/lmignon

**Comment by pedrobaeza** [comment]
Well, tell me when you finish with all, and I'll merge this with a bit of squashing.

**Comment by lmignon** [comment]
@pedrobaeza I've reworked the commit history. This PR is ready to merge. I'll rebase the others based on this one..

**Pull request merged**
